projects/ziti-console-lib/src/lib/features/projectable-forms/service/service-form.component.ts

```diff
@@ -0,0 +1,561 @@
+    let selectedRouter;
+    if (this.selectedConfigId) {
+      this.routers.forEach((router) => {
+        if (this.selectedRouterId === router.id) {
+          selectedRouter = router;
+        }
+      });
+    }
+    this.selectedRouter = selectedRouter;
+  }
+
+  addTerminator(item) {
+    let termAdded = false;
+    this.addedTerminators.forEach((termName) => {
+      if (item.name === termName) {
+        termAdded = true;
+      }
+    });
+    if (!termAdded) {
+      this.addedTerminators.push(item.name);
+    }
+  }
+
+  bindingTypeChanged(event?: any) {
+
+  }
+
+  terminatorProtocolChanged(event?: any) {
+
+  }
+
+  get showConfigData() {
+    return this.selectedConfigId === 'add-new';
+  }
+
+  get showAttachButton() {
+    return !isEmpty(this.selectedConfigId);
+  }
+
+  getAssociatedIdentities() {
+    this.zitiService.getSubdata('services', this.formData.id, 'identities').then((result: any) => {
+      this.associatedIdentities = result.data;
+      this.associatedIdentityNames = this.associatedIdentities.map((svc) => {
+        return svc.name;
+      });
+    });
+  }
+
+  headerActionRequested(action) {
+    switch(action.name) {
+      case 'save':
+        this.save();
+        break;
+      case 'close':
+        this.closeModal(true);
+        break;
+      case 'toggle-view':
+        this.formView = action.data;
+        break;
+    }
+  }
+
+  async save(event?) {
+    const isValid = this.validate();
+    const isExtValid = await this.extService.validateData();
+    if(!isValid || !isExtValid) {
+      return;
+    }
+
+    this.isLoading = true;
+    this.svc.save(this.formData).then((result) => {
+      if (!isEmpty(result.id)) {
+        this.formData = result;
+        this.initData = this.formData;
+      }
+    }).finally(() => {
+      this.isLoading = false;
+    });
+  }
+
+  validate() {
+    this.errors = {};
+    if (isEmpty(this.formData.name)) {
+      this.errors['name'] = true;
+    }
+    return isEmpty(this.errors);
+  }
+
+  validateConfig() {
+    this.configErrors = {};
+    if (isEmpty(this.newConfigName)) {
+      this.configErrors['name'] = true;
+    }
+    return isEmpty(this.configErrors);
+  }
+
+  toggleEncryptionRequired() {
+    this.formData.encryptionRequired = !this.formData.encryptionRequired;
+  }
+
+  closeModal(refresh = true, ignoreChanges = false): void {
+    if (!ignoreChanges && this._dataChange) {
+      const confirmed = confirm('You have unsaved changes. Do you want to leave this page and discard your changes or stay on this page?');
+      if (!confirmed) {
+        return;
+      }
+    }
+    this.close.emit({refresh: refresh});
+  }
+
+  clear(): void {
+  }
+
+  _dataChange = false;
+  watchData() {
+    delay(() => {
+      const dataChange = !isEqual(this.initData, this.formData);
+      if (dataChange !== this._dataChange) {
+        this.dataChange.emit(dataChange);
+      }
+      this._dataChange = dataChange;
+      this.watchData();
+    }, 100);
```

Comment:
this seems like the sort of thing that should be a component or at least a function abstracted away, no? also this is polling? :/ Is it making a request every 100 ms?